### PR TITLE
feat: Support certain providers to customize the base URL.

### PR DESCRIPTION
**Pull request by Emt-lin**
related: #654

Support certain providers to customize the base URL.
 - Refactor the method of geting model config.
 - Without introducing opeai package, CORS and other feature is suported.

### CODE REVIEW

File: src/LLMProviders/embeddingManager.ts [hunk 1]
```diff
@@ -19,11 +32,20 @@ export default class EmbeddingManager {
     string,
     {
       hasApiKey: boolean;
-      EmbeddingConstructor: new (config: any) => Embeddings;
+      EmbeddingConstructor: EmbeddingConstructorType;
       vendor: string;
     }
   >;
 
+  private readonly providerAipKeyMap: Record<EmbeddingModelProviders, () => string> = {
```
Comment (logancyang) [inline src/LLMProviders/embeddingManager.ts:40]: nit: Typo, `Api`

File: src/langchainWrappers.ts [hunk 2]
```diff
@@ -1,9 +1,8 @@
...
 import { AnthropicInput, ChatAnthropic } from "@langchain/anthropic";
-import { ChatOpenAI } from "@langchain/openai";
-import { OpenAIEmbeddings } from "@langchain/openai";
-import { requestUrl } from "obsidian";
+import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
 import OpenAI from "openai";
+import { safeFetch } from "@/utils";
 
 // Migrated to OpenAI v4 client from v3: https://github.com/openai/openai-node/discussions/217
 export class ProxyChatOpenAI extends ChatOpenAI {
```
Comment (logancyang) [inline src/langchainWrappers.ts:8]: Does this mean we can get rid of `ProxyChatOpenAI`?

**Comment by logancyang** [comment]
@Emt-lin Nice, thanks for making the new PRs! Some nits but generally LGTM. 

I'll go ahead and merge it and fix the nits myself. Do you use any AI-assistant IDE like Cursor? It will be helpful to catch typos and problems using AI first before submitting the PR to have an extra level of check.

File: src/langchainWrappers.ts (hunk 2 above)
Comment (Emt-lin) [reply src/langchainWrappers.ts:8]: @logancyang yeah, perhaps I usually don't want to delete redundant code. I don't know why -_-

**Comment by Emt-lin** [comment]
> I'll go ahead and merge it and fix the nits myself. Do you use any AI-assistant IDE like Cursor? 

I usually use the Intellij IDEA,  and sometimes I use the cursor. 

> It will be helpful to catch typos and problems using AI first before submitting the PR to have an extra level of check.

OK

**Pull request merged**